### PR TITLE
Login using email links

### PR DESCRIPTION
**Pull request by valo**
Add GraphQL APIs for requesting an email with a login link and for validating a login token and email.

The login links are valid for 24 hours.

The emails are being sent using mandrill, using a predefined template there. We need to get the mandrill transactional emails enabled and test this locally.

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -199,6 +200,10 @@ config :ex_aws,
 
 config :sanbase, Sanbase.Etherbi, url: {:system, "ETHERBI_URL"}
 
+config :sanbase, Sanbase.MandrillApi,
+  apikey: {:system, "MANDRILL_APIKEY"},
+  from_email: {:system, "MANDRILL_FROM_EMAIL", "admin@santiment.net"}
```
Comment (IvanIvanoff) [inline config/config.exs]: Wouldn't it be better if we had `login@santiment.net` or something like that?

Comment (valo) [reply config/config.exs]: For now we haven't registered such email address and I haven't explored how email forwarding works exactly. Is there any particular reason we might want to change it now?

File: lib/sanbase/auth/user.ex [hunk 2]
```diff
@@ -63,4 +80,58 @@ defmodule Sanbase.Auth.User do
...
+    cond do
+      user.email_token != token ->
+        false
+
+      Timex.diff(Timex.now(), user.email_token_generated_at, :minutes) >
+          @login_email_valid_minutes ->
+        false
+
+      user.email_token_validated_at == nil ->
+        true
```
Comment (IvanIvanoff) [inline lib/sanbase/auth/user.ex]: Should return `false` if the token is not validated?

Comment (valo) [reply lib/sanbase/auth/user.ex]: If the token is never validated and everything from the above is not true, we should consider it valid.

File: config/config.exs (hunk 1 above)
Comment (IvanIvanoff) [reply config/config.exs]: No, just aesthetic reasons

File: lib/sanbase/mandrill_api.ex [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      template_name: template,
+      template_content: [],
+      message: %{
+        to: [
+          %{
+            email: recepient
+          }
+        ],
+        global_merge_vars: build_global_merge_vars(variables),
+        from_email: Config.module_get(__MODULE__, :from_email),
```
Comment (IvanIvanoff) [inline lib/sanbase/mandrill_api.ex]: Fetching from the same module is shorter with `Config.get(:from_email)`.

File: lib/sanbase_web/endpoint.ex [hunk 4]
```diff
@@ -58,4 +60,12 @@ defmodule SanbaseWeb.Endpoint do
       {:ok, config}
     end
   end
+
+  def website_url() do
+    Config.module_get(__MODULE__, :website_url)
```
Comment (IvanIvanoff) [inline lib/sanbase_web/endpoint.ex]: `Config.get(:website_url)`

File: lib/sanbase_web/graphql/resolvers/account_resolver.ex [hunk 5]
```diff
@@ -39,6 +39,27 @@ defmodule SanbaseWeb.Graphql.Resolvers.AccountResolver do
...
+  end
+
+  def email_login_verify(%{token: token, email: email}, _resolution) do
+    with {:ok, user} <- User.find_or_insert_by_email(email),
+         true <- User.email_token_valid?(user, token),
+         {:ok, token, _claims} <- SanbaseWeb.Guardian.encode_and_sign(user, %{salt: user.salt}),
+         {:ok, user} <- User.mark_email_token_as_validated(user) do
+      {:ok, %{user: user, token: token}}
+    else
+      _ -> {:error, :login_failed}
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/account_resolver.ex]: Unify the error messages. IMO the second argument being just a string is most readable (use `message` when you want to specify `details`, too)

File: test/sanbase/auth/user_test.exs [hunk 6]
```diff
@@ -62,4 +62,71 @@ defmodule Sanbase.Auth.UserTest do
...
+
+  test "update_email_token updates the email_token and the email_token_generated_at" do
+    user =
+      %User{salt: User.generate_salt()}
+      |> Repo.insert!()
+
+    {:ok, user} = User.update_email_token(user)
+
+    assert user.email_token != nil
+    assert Timex.diff(Timex.now(), user.email_token_generated_at, :seconds) == 0
```
Comment (IvanIvanoff) [inline test/sanbase/auth/user_test.exs]: Could this fail if the VM decides to switch off this process and return after more than 1 second?

Comment (valo) [reply test/sanbase/auth/user_test.exs]: Switching off the process for more than 1 sec seems very unlikely to me. If we start getting random failures in the builds we can change this.

**Pull request merged**